lib/auth/crownjewel/object.go

```diff
@@ -0,0 +1,105 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package crownjewel
+
+import (
+	"github.com/gravitational/trace"
+
+	crownjewelv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/crownjewel/v1"
+	headerv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
+)
+
+// NewCrownJewel creates a new CrownJewel object.
+// It validates the object before returning it.
+func NewCrownJewel(name string, spec *crownjewelv1.CrownJewelSpec) (*crownjewelv1.CrownJewel, error) {
+	cj := &crownjewelv1.CrownJewel{
+		Metadata: &headerv1.Metadata{
+			Name: name,
+		},
+		Spec: spec,
+	}
+
+	if err := ValidateCrownJewel(cj); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return cj, nil
+}
+
+// ValidateCrownJewel validates the CrownJewel object.
+// Required fields:
+//   - Metadata.Name
+//   - Spec.TeleportMatchers or Spec.AwsMatchers
+//   - Matcher.Name or Matcher.Labels
+func ValidateCrownJewel(jewel *crownjewelv1.CrownJewel) error {
+	switch {
+	case jewel == nil:
```

Comment:
for this and the later parts: a `nil` pointer protobuf message is a valid read-only message that has all fields zeroed; should we start using getters whenever possible? we've kind of avoided that for gogoproto, which doesn't always generate getters and can store submessages as structs, but that's not how the convention goes for protobuf